# Sistemas Dinamicos

Un sistema dinamico es un sistema cuyo <b>estado</b> esta unicamente especificado por un conjunto de variables y su <b>comportamtiento</b> esta definido por un conjunto de reglas.

<b>Ejemplos</b>:
- Crecimiento de poblacion.
- Pendulo.
- Movimiento de planetas.
- Sistema financiero.

### Sistema Discreto

Este tipo de modelo es llamado <b>ecuacion de diferencias</b>

$$
x_t = F(x_{t-1}, t)
$$

### Sistema Continuo

Este tipo de modelo es llamado <b>ecuacion diferenncial</b>

$$
\frac{dx}{dt} = F(x,t)
$$

<hr>

## Clasificaciones

### Lineal o no lineal

Un sistema es no lineal cuando posee variables de estado multiplicandose.

<b>Ejemplo</b> lineal:

$$
x_t = 2 x_{t-1}
$$

<b>Ejemplo</b> no lineal:

$$
x_t = 2 x_{t-1} y_{t-1}
$$

### Orden

Un sistema puede ser, de primer orden, segundo, tercero, etc. Dependiendo de si para calcular el estado actual $t$ dependende de uno, dos, tres, estados anteriores.

<b>Ejemplo</b> primer orden:

$$
x_t = 2 t_{t-1}
$$

<b>Ejemplo</b> segundo orden:

$$
x_t = 2 x_{t-1} + x_{t-2}
$$

### Autonomo o no autonomo

Un sistema es no autonomo, cuando aparece la variable tiempo de forma explicita.

<b>Ejemplo</b> autonomo:

$$
x_t = 2 x_{t-1}
$$

<b>Ejemplo</b> no autonomo:

$$
x_t = 2 t x_{t-1}
$$

<hr>

## Punto de Equilibrio

Son los puntos donde el sistema deja de pregresar, cuando ocurre que el estado siguiente es el mismo que el anterior. Matematicamente en un sistema discreto esto se expresa de la siguiente manera:

$$
x_{t + 1} = x_t
$$

## Espacio de Fases

El espacio de fases es el conjunto de los diferentes estados posibles, cualquier estado en el sistema se representa por un punto en el estado de fases.

Un sistema dinamico evoluciona en el tiempo como una <b>trayectoria</b> en el espacio de fases.

### Comportamiento de las trayectorias

- Diverger al infinito.
- Converger a un <b>punto</b> fijo estable.
- Oscilar al rededor de un <b>punto</b>.

A estos puntos de los ultimos dos items se los denomina <b>atractor</b>.

<b>Cuenca del atractor</b>: conjunto de puntos en el espacio de fases que llevan a un atractor.

<img src="imagenes/comportamiento_trayectorias.png" width="500" />

## Comportamiento en Sistemas Lineales

Podemos lograr conocer el comportamiento asintotico, mediante el estudio de los autovalores.

Si el autovalor dominante $\lambda_d$ en modulo es mayor a 1. Entonces el sistema <b>diverge</b>.

<b>Nota</b>: tienen un unico punto fijo. 

#### Ejemplo

Teniendo el siguiente <b>Sistema Dinamico</b>:

$$
x_t = x_{t - 1} - y_{t - 1}
$$

$$
y_t = -x_{t - 1} - 3y_{t - 1} + z_{t - 1}
$$

$$
z_t = y_{t - 1} + z_{t - 1}
$$

Construimos la <b>matriz</b>

$$
A=
\begin{bmatrix}
1  & -1 & 0 \\
-1 & -3 & 1 \\
0  &  1 & 1 \\
\end{bmatrix}
$$

Calculamos los <b>autovalores</b>

$$
det(A - \lambda I) = 0
$$

In [3]:
import numpy as np
m = np.matrix('1 -1 0 ; -1 -3 1 ; 0 1 1')
m

matrix([[ 1, -1,  0],
        [-1, -3,  1],
        [ 0,  1,  1]])

In [7]:
autovalores = np.linalg.eigvals(m)

Nos quedamos con el autovalor, <b>dominante</b> (en modulo)

In [9]:
abs(autovalores[0])

3.4494897427831788

Como es mayor a 1, entonces el sistema <b>diverge</b>

## Comportamiento Local en Sistemas No Lineales

Cuando el sistema es no lineal, la matematica para encontrar el comportamiento cambia, debemos encontrar los puntos fijos. Crear la matriz jacobiana y evaluar estos puntos en ella.

<b>Observacion</b>: lo que se estudia es el comportamiento <b>local</b>, es decir en un entorno reducido al rededor del punto.

<img src="imagenes/estudio_local_punto_fijo_sistema_no_lineal.png" />

Entonces:

1) Encontramos los puntos de equilibrio plenteando $x_{t-1} = x_t = x$.

2) Calculamos la <b>matriz Jacobiana</b>.

Ejemplo caso de dos ecuaciones.

$$
J =
\begin{bmatrix}
\frac{df}{dx} & \frac{df}{dy} \\
\frac{dg}{dx} & \frac{dg}{dy} \\
\end{bmatrix} 
$$

3) Evaluamos los puntos de equilibrio en la matriz jacobiana.

4) Calculamos y analizamos los autovalores. Y los clasificamos mediante la siguiente tabla:

### Discreto 

<img src="imagenes/comportamiento_no_lineal_discreto.png" />

### Continuo

<img src="imagenes/comportamiento_no_lineal_continuo.png" />

### Ejemplo caso Discreto

$$
x_t = x_{t - 1} y_{t - 1}
$$

$$
y_t = y_{t - 1} (x_{t - 1} - 1) 
$$

<b>Observacion</b>: notar que no es un sistema lineal. Hay dos variables que se estan multiplicando.

1) Buscamos los <b>puntos de equilibrio</b>:

$$
x_t = x_{t - 1} = x
$$

$$
y_t = y_{t - 1} = y
$$

Remplazando en el sistema:

$$
x = x y
$$

$$
y = y(x - 1)
$$

(0, 0) es una solucion

(1, 1) es la otra

Por lo tanto los dos puntos de equilibrio son: $(0,0)$ y $(1,1)$

2) Calculamos la matriz <b>jacobiana</b>

$$
f(x,y) = xy
$$
$$
g(x,y) = y(x - 1) = yx - y
$$

$$
J(x,y) =
\begin{bmatrix}
\frac{df}{dx} & \frac{df}{dy} \\
\frac{dg}{dx} & \frac{dg}{dy} \\
\end{bmatrix}
$$

$$
J(x,y) =
\begin{bmatrix}
y & x \\
y & x - 1 \\
\end{bmatrix}
$$

3) Evaluamos el jacobiano en en los puntos de equilibrio, y calculamos los autovalores. Y clasificamos al punto.

Estudio del $(0,0)$

$$
J(0,0) =
\begin{bmatrix}
0 & 0 \\
0 & -1 \\
\end{bmatrix}
$$

In [10]:
m = np.matrix('0 0 ; 0 -1')
m

matrix([[ 0,  0],
        [ 0, -1]])

In [12]:
autovalores = np.linalg.eigvals(m)
autovalores

array([ 0., -1.])

De los dos autovalores el dominante es $$|\lambda_d| = 1$$

Por lo tanto el comportamiento local de este punto es el de un <b>Lyapunov stable</b>, <b>neutral point</b>

<img src="imagenes/comportamiento_neutral_point.png" />

Estudio del $(1,1)$

$$
J(1,1) =
\begin{bmatrix}
1 & 1 \\
1 & 0 \\
\end{bmatrix}
$$

In [13]:
m = np.matrix('1 1 ; 1 0')
m

matrix([[1, 1],
        [1, 0]])

In [14]:
autovalores = np.linalg.eigvals(m)
autovalores

array([ 1.61803399, -0.61803399])

De los dos autovalores el dominante es $$|\lambda_d| = 1.618$$

Por lo tanto el comportamiento local de este punto es el de un <b>Unstable</b>, <b>saddle point</b>

<img src="imagenes/comportamiento_saddle_point.png" />

### Ejemplo caso Continuo

$$
\frac{dx}{dt} = x - xy
$$

$$
\frac{dy}{dt} = -y + xy
$$

1) Buscamos los <b>puntos de equilibrio</b>:

Para esto se plantea:

$$
\frac{dx}{dt} = 0
$$

$$
\frac{dy}{dt} = 0
$$

$$
x = xy
$$

$$
y = xy
$$

Una solucion es $(0,0)$, y la otra es $(1,1)$

2) Calculamos la matriz <b>jacobiana</b>

$$
f(x,y) = x - xy
$$
$$
g(x,y) = -y + xy
$$

$$
J(x,y) =
\begin{bmatrix}
\frac{df}{dx} & \frac{df}{dy} \\
\frac{dg}{dx} & \frac{dg}{dy} \\
\end{bmatrix}
$$